extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp

```diff
@@ -0,0 +1,310 @@
+    }
+
+    void run() {
+      configureSecurity();
+
+      createServerSocket();
+
+      verifyTLSServerSocketExclusiveCompatibilityWithTLSv1_2();
+
+      shutdownServerSocket();
+    }
+
+ protected:
+    void configureSecurity() {
+      host_ = org::apache::nifi::minifi::io::Socket::getMyHostName();
+      port_ = "38778";
+      if (!key_dir_.empty()) {
+        configuration_->set(minifi::Configure::nifi_remote_input_secure, "true");
+        configuration_->set(minifi::Configure::nifi_security_client_certificate, key_dir_ + "cn.crt.pem");
+        configuration_->set(minifi::Configure::nifi_security_client_private_key, key_dir_ + "cn.ckey.pem");
+        configuration_->set(minifi::Configure::nifi_security_client_pass_phrase, key_dir_ + "cn.pass");
+        configuration_->set(minifi::Configure::nifi_security_client_ca_certificate, key_dir_ + "nifi-cert.pem");
+        configuration_->set(minifi::Configure::nifi_default_directory, key_dir_);
+      }
+    }
+
+    void createServerSocket() {
+      const auto socket_context = std::make_shared<org::apache::nifi::minifi::io::TLSContext>(configuration_);
+      server_socket_ = std::make_shared<org::apache::nifi::minifi::io::TLSServerSocket>(socket_context, host_, std::stoi(port_), 3);
+      assert(0 == server_socket_->initialize());
+
+      is_running_ = true;
+      check_ = [this]() -> bool {
+        return is_running_;
+      };
+      handler_ = [this](std::vector<uint8_t> *bytes_written, int *size) {
+        std::string contents = "hello world";
+        *bytes_written = {contents.begin(), contents.end()};
+        bytes_written->push_back(0);
+        *size = bytes_written->size();
+        return *size;
+      };
+      server_socket_->registerCallback(check_, handler_, std::chrono::milliseconds(50));
+    }
+
+    void verifyTLSServerSocketExclusiveCompatibilityWithTLSv1_2() {
+      verifyTLSProtocolIncompatibility<SimpleSSLTestClientTLSv1>();
+      verifyTLSProtocolIncompatibility<SimpleSSLTestClientTLSv1_1>();
+      verifyTLSProtocolCompatibility<SimpleSSLTestClientTLSv1_2>();
+    }
+
+    template <class TLSTestClient>
+    void verifyTLSProtocolIncompatibility() {
+      verifyTLSProtocolCompatibility<TLSTestClient>(false);
+    }
+
+    template <class TLSTestClient>
+    void verifyTLSProtocolCompatibility() {
+      verifyTLSProtocolCompatibility<TLSTestClient>(true);
+    }
+
+    template <class TLSTestClient>
+    void verifyTLSProtocolCompatibility(bool should_be_compatible) {
+      TLSTestClient client(host_, port_);
+      assert(client.canConnect() == should_be_compatible);
+    }
+
+    void shutdownServerSocket() {
+      is_running_ = false;
+    }
+
+    std::function<bool()> check_;
+    std::function<int(std::vector<uint8_t> *bytes_written, int *size)> handler_;
+    std::atomic<bool> is_running_;
+    std::shared_ptr<org::apache::nifi::minifi::io::TLSServerSocket> server_socket_;
```

Comment:
we could use a `unique_ptr` for the `server_socket_` (to make it clear that it is getting destroyed at the end of this object) and maybe even call an explicit `reset` on it in the destructor of `TLSServerSocketSupportedProtocolsTest ` (since currently the correctness hinges on the order of `server_socket_` and `is_running_`)

Reply:
Made it unique to be more clear, calling reset however seemed little excessive since unique_ptr destructor will be called and that achieves the same thing. I did however changed the destructor to set is_running_ to false. This will basically achieve that the external check function inside the socket will return false, after that the unique_ptr destructor will destroy the socket which will set its own inside check to false and it'll wait for the thread to finish.

Reply:
well if in the future somebody switches up `is_running_` and `server_socket_`, `is_running_` will get destroyed first, and then the still running `server_socket_` could accidentally access the then destroyed `is_running_` (and setting `is_running_` in the destructor does not seem to mitigate this), but you are right that it is currently not an issue, maybe a comment then, that the order of these fields is important?

Reply:
I see your point, you are right, an explicit reset call in the destructor before the destruction of the members seems the best way. fixed